Keyboard

SirEndii opened the pull request:
* **Please check if the PR fulfills these requirements**
- [x] The commit message are well described
- [ ] Docs have been added / updated
- [x] All changes have fully been tested

* **What kind of change does this PR introduce?** (Bug fix, feature, ...)
Feature

* **What is the new behavior (if this is a feature change)?**
Added the keyboard item. An item which can be linked to a computer. A user can now right click the item to unlock their mouse, they can now freely type while the typed characters are send to the linked computer. Will also be a module for the smart glasses.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their scripts due to this PR?)
No

* **Other information**:

SirEndii committed:
[IntelligenceModding/Advanced-Peripherals-Features#59] Keyboard item and simple screen to unlock the mouse

SirEndii committed:
Remove custom indentation

SirEndii committed:
Binding/Unbinding for the keyboard

SirEndii committed:
Use translatable components and add name and description for the keyboard
Also fixed colored messages for the memory card and cleaned up the translation keys

SirEndii committed:
Removed owner tag from the memory card

SirEndii committed:
Key press/Char typed sync to the computer logic

SirEndii commented:
So the keyboard works now
I can bound it to a computer, right click it in the air and start typing. Every key stroke, key presses and releases are synced to the terminal

I now want to remove JEI from the keyboard screen and place a little info text at the top

The last thing I need to do is to implement this as a smart glasses module

SirEndii committed:
The keyboard is now a working module, it currently uses the hotkey module keybind. I will see what I will do with the hotkey module or if I will create a new Keybind
I've created two new nbt tags to bind the module to the glasses to sync that to the container, I am not sure if there is maybe a better way, but it works

SirEndii committed:
Mouse events for the keyboard module

SirEndii committed:
Fixed an issue where the keyboard does not bound to the glasses when it was bound to different glasses before

SirEndii committed:
Render a little info text when the player is using the keyboard

SirEndii committed:
Preven JEI/REI/EMI from rendering in the keyboard screen

SirEndii commented:
I think everything is done now. Mouse and keyboard events are working, I got the little info text rendered, JEI does not render anymore and everything works fine as a glasses module

zyxkad commented on src/main/java/de/srendi/advancedperipherals/client/screens/KeyboardScreen.java:
```diff
@@ -0,0 +1,218 @@
+        float textX = (screenWidth / 2f - minecraft.font.width(text) * scale / 2f) / scale;
+        minecraft.font.drawShadow(poseStack, text, textX, 1, 0xFFFFFF);
+
+        // Prevents JEI/REI/EMI from rendering. Maybe not the best way, but it works for now.
```
Would you mind explain a bit what are you doing here to prevent them render? Or I mean I want to know if this won't disable other mods that decides to constant render something on the screen, such as inventory hud+

SirEndii replied:
I have actually no clue if it also disables other ui addon stuff
So I just render an empty string on the far lower right and jei thinks it does not have space to render
There is maybe a better way, maybe using the JEI API

SirEndii replied:
I simplified that
I now just define a huge screen image which tells minecraft that the actual screen which we don't need to render is huge
That prevents jei from rendering

zyxkad commented on src/main/java/de/srendi/advancedperipherals/client/screens/KeyboardScreen.java:
```diff
@@ -0,0 +1,218 @@
+
+    @Override
+    public boolean charTyped(char ch, int modifiers) {
+        if (ch >= 32 && ch <= 126 || ch >= 160 && ch <= 255) { // printable chars in byte range
```
ASCII char has a maximum value of `0x7f` (127) btw, not `0xff` or 255, above 0x7f will be considers as a two or more bytes UTF8

SirEndii replied:
Everything below charTyped is borrowed from CCs screen itself, so I don't know why squid did what 😅

zyxkad commented on src/main/java/de/srendi/advancedperipherals/client/screens/KeyboardScreen.java:
```diff
@@ -0,0 +1,218 @@
+                        clipboard = SharedConstants.filterText(clipboard);
+                        if (!clipboard.isEmpty()) {
+                            // Clip to 512 characters and queue the event
+                            if (clipboard.length() > 512) clipboard = clipboard.substring(0, 512);
```
I guess we better clip before check anything should we not?

zyxkad reviewed (commented):
Overall looks good, just have some detail need to deal with

SirEndii commented:
Sorry to annoy you @SquidDev, just asking you because of https://github.com/IntelligenceModding/AdvancedPeripherals/pull/650#issuecomment-2254387033

I used a piece of the TerminalWidget implementation for our keyboard screen. Just asking if this is okay
See
https://github.com/IntelligenceModding/AdvancedPeripherals/pull/659/files#diff-6374267185184dd9bdd6868cb47752c688a105088d0063d0c5bf7992b4e10ec9R19

And maybe you can answer the two comments related to that from zyxkad

SquidDev commented:
It might be easier to copy what [`NoTermComputerScreen` does](https://github.com/cc-tweaked/CC-Tweaked/blob/c271ed7c7f9c9e3e927a0939fafb20690b8d40f3/projects/common/src/client/java/dan200/computercraft/client/gui/NoTermComputerScreen.java), and just create a `Terminal` but never render it. Plethora's keyboard UI [actually just subclasses `NoTermComputerScreen` directly](https://github.com/SwitchCraftCC/Plethora-Fabric/blob/1.20.1/src/main/java/io/sc3/plethora/gameplay/modules/keyboard/KeyboardComputerScreen.kt).

SirEndii commented:
> It might be easier to copy what [`NoTermComputerScreen` does](https://github.com/cc-tweaked/CC-Tweaked/blob/c271ed7c7f9c9e3e927a0939fafb20690b8d40f3/projects/common/src/client/java/dan200/computercraft/client/gui/NoTermComputerScreen.java), and just create a `Terminal` but never render it. Plethora's keyboard UI [actually just subclasses `NoTermComputerScreen` directly](https://github.com/SwitchCraftCC/Plethora-Fabric/blob/1.20.1/src/main/java/io/sc3/plethora/gameplay/modules/keyboard/KeyboardComputerScreen.kt).

Thank you, I didn't see that
That would be simpler

SirEndii committed:
Merge branch 'dev/0.8' into feat/keyboard

SirEndii committed:
checkstyle

SirEndii committed:
Simplify KeyboardScreen

zyxkad commented on build.gradle:
```diff
@@ -324,11 +324,12 @@ dependencies {
     compileOnly fg.deobf("com.ldtteam:blockui:${blockui_version}")
     // IMPORTANT. This should be removed/commented when running `runData`
-    runtimeOnly fg.deobf("com.ldtteam:minecolonies:${minecolonies_version}")
+    /*runtimeOnly fg.deobf("com.ldtteam:minecolonies:${minecolonies_version}")
```
I pretend I didn't see those

(although I think you should merge from dev/0.8)

zyxkad commented on src/main/java/de/srendi/advancedperipherals/client/ClientRegistry.java:
```diff
@@ -43,6 +44,7 @@ public static void registerModels(ModelEvent.RegisterAdditional event) {
     public static void onClientSetup(FMLClientSetupEvent event) {
         MenuScreens.register(APContainerTypes.INVENTORY_MANAGER_CONTAINER.get(), InventoryManagerScreen::new);
         MenuScreens.register(APContainerTypes.SMART_GLASSES_CONTAINER.get(), SmartGlassesScreen::new);
+        MenuScreens.register(APContainerTypes.KEYBOARD_CONTAINER.get(), KeyboardScreen::new);
```
keep things sorted!
```suggestion
        MenuScreens.register(APContainerTypes.INVENTORY_MANAGER_CONTAINER.get(), InventoryManagerScreen::new);
        MenuScreens.register(APContainerTypes.KEYBOARD_CONTAINER.get(), KeyboardScreen::new);
        MenuScreens.register(APContainerTypes.SMART_GLASSES_CONTAINER.get(), SmartGlassesScreen::new);
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/client/screens/KeyboardScreen.java:
```diff
@@ -0,0 +1,126 @@
+ * <p>
+ * We just create a terminal which is used to forward all the key presses and mouse clicks but we don't render it.
+ */
+public class KeyboardScreen extends BaseScreen<KeyboardContainer> {
```
If you do not extend from `AbstractContainerScreen`, but inherit `Screen` or `GuiComponent`, then jade should not render on it

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/container/KeyboardContainer.java:
```diff
@@ -0,0 +1,69 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class KeyboardContainer extends BaseContainer implements ComputerMenu {
```
And so we don't need Keyboard*Container* anymore

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/items/MemoryCardItem.java:
```diff
@@ -30,9 +30,8 @@ public boolean isEnabled() {
-        if (data.contains("ownerId") && data.contains("owner")) {
-            tooltip.add(EnumColor.buildTextComponent(Component.translatable("item.advancedperipherals.tooltip.memory_card.bound", data.getString("owner"))));
+        if (data.contains("ownerId")) {
+            tooltip.add(EnumColor.buildTextComponent(Component.translatable("item.advancedperipherals.tooltip.binding.boundto", data.getString("ownerId"))));
```
alr we are not showing ownerId here, we still want show the owner's name, so we need an extra step to turn the id into `GameProfile`, then get the player's cached name

SirEndii replied:
ah yeah
I can use the new cache thing there I made

SirEndii committed:
Sort ClientRegistry.java

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
Merge branch 'dev/0.8' into feat/keyboard

# Conflicts:
#	build.gradle
#	src/generated/resources/.cache/67cce32b1c3cbbcb1f646605f4914e3f196986c2
#	src/generated/resources/.cache/c622617f6fabf890a00b9275cd5f643584a8a2c8
#	src/generated/resources/assets/advancedperipherals/lang/en_us.json
#	src/main/java/de/srendi/advancedperipherals/common/data/EnUsLanguageProvider.java
#	src/main/java/de/srendi/advancedperipherals/common/setup/APItems.java

SirEndii committed:
Add language entries

zyxkad committed:
Merge branch 'dev/0.8' of https://github.com/IntelligenceModding/AdvancedPeripherals into player-rotate-handle-feature

zyxkad committed:
make keyboard based on computer ID instead of position

zyxkad committed:
run runData

zyxkad committed:
implement keyboard mouse capture mode
- fix keyboard container will keep reopen whenever holding G
- rename glassesKeyPressed event to glasses_key_pressed
- fix JEI will shown on keyboad container screen

zyxkad committed:
fix styles

SirEndii committed:
Merge pull request #719 from zyxkad/player-rotate-handle-feature

Player rotate handle feature

zyxkad commented on src/main/java/de/srendi/advancedperipherals/client/screens/KeyboardScreen.java:
```diff
@@ -0,0 +1,239 @@
+    private MouseState mouseState = MouseState.RELEASED;
+    private boolean captureMouse;
+    private boolean regrabingMouse;
+    private byte[] lastPosLock = new byte[0];
```
```suggestion
    private final byte[] lastPosLock = new byte[0];
```
I forgot make this final so it confused me a bit

zyxkad committed:
make lock final

Signed-off-by: Kevin Z <zyxkad@gmail.com>

zyxkad committed:
fix keyboard does not work in multiplayer

zyxkad committed:
fix keyboard_close fires with delay

zyxkad committed:
remove unused import

SirEndii committed:
Merge pull request #751 from zyxkad/feat/keyboard

Make keyboard works in multiplayer

zyxkad committed:
add license to KeyboardScreen and some code cleanup

zyxkad committed:
Merge branch 'dev/0.8' of https://github.com/IntelligenceModding/AdvancedPeripherals into feat/keyboard

zyxkad commented:
Force merge cause 1.19 is not going to be supported.
